### PR TITLE
fix(config): keep installed versions eligible with release age

### PR DESCRIPTION
**Pull request by jdx**
## Summary

- keep already-installed fuzzy matches eligible when `minimum_release_age` is configured
- reserve date-aware filtering of installed matches for lockfile generation, which intentionally discovers versions from release metadata
- document the installed-version semantics and add regression coverage for local resolution and prune

## Root cause

An explicit `minimum_release_age` cutoff globally disabled installed-version fast paths. As a result, ordinary online resolution could select an older age-eligible remote release while offline/prefer-offline resolution and prune selected the latest installed release. This made the same `latest` request resolve inconsistently and made prune appear to remove an active version.

`minimum_release_age` is an installation/discovery gate: an explicitly installed version remains eligible. Commands that intentionally discover versions (`upgrade`, install resolution, `latest`, `ls-remote`, and date-aware lock generation) continue to apply the cutoff to remote candidates.

Addresses discussion #10944.

## Validation

- `mise run lint-fix`
- `mise run test:unit` (1,668 passed)
- `mise run test:e2e test_minimum_release_age_installed_resolution`
- `mise run test:e2e test_install_prefer_offline_no_refresh test_which_no_remote_fetch test_which_prefer_offline`
- `mise run test:e2e e2e/cli/test_lock` (target test passed)
- `mise run test:e2e test_npm_install_before`

---

> [!NOTE]
> **Medium Risk**
> Changes central version resolution in `tool_version.rs`, which affects most commands that resolve tools; behavior is narrowed with explicit opt-in for lockfile discovery and regression tests, but edge cases around fuzzy specs and release age remain possible.
> 
> **Overview**
> Fixes inconsistent behavior where an explicit **`minimum_release_age`** (or built-in default cutoff) could skip installed-version fast paths, forcing remote lookups for fuzzy specs like **`latest`** and making online resolution, **`mise which`**, and **`mise prune`** disagree about the active version.
> 
> Resolution now treats release age as a **remote discovery** gate: **`ResolveOptions`** adds **`filter_installed_versions_by_release_date`** (default **off**). Ordinary flows (**`install`**, **`use`**, **`upgrade`**, hook-env-style resolution) leave it **false**, so already-installed fuzzy matches stay eligible without re-checking release metadata. **`mise lock`** sets it **true** and may rebuild the toolset when a release-age policy is configured, so lock generation can still pick age-eligible remote versions over a newer installed match.
> 
> Docs in **`docs/security.md`** spell out that installed versions are not made inactive by the setting. E2E coverage adds **`test_minimum_release_age_installed_resolution`** and adjusts lock / which regression tests.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 02b677e0a4c1e6a8c44df292d1ae69703647a312. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

- **Bug Fixes**
  - Already-installed tool versions remain eligible during resolution even when a minimum release age is configured.
  - `mise install`, `mise use`, and `mise upgrade` now avoid release-age filtering for installed matches and reduce unnecessary remote lookups.
  - `mise lock` continues to apply minimum release age filtering when selecting versions.
- **Documentation**
  - Updated security documentation to clarify that minimum release age affects only remote version selection, not installed versions.
- **Tests**
  - Added/updated end-to-end coverage for minimum-release-age installed-resolution behavior and offline/lookup preference scenarios.

### CODE REVIEW

File: src/cli/lock.rs [hunk 1]
```diff
@@ -87,6 +87,7 @@ impl Lock {
         let before_date = self.get_before_date()?;
         let lock_resolve_options = ResolveOptions {
             before_date,
+            filter_installed_versions_by_release_date: true,
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/cli/lock.rs:90]: Fixed in d917c6b60. Non-monorepo lock generation now rebuilds the toolset with the lock-specific resolve options even when the CLI cutoff is unset, so global and per-tool cutoffs can filter an installed fuzzy match. The lock regression now creates the newer installed version before exercising both cutoff sources. I also kept the built-in default on the installed fast path to preserve the #10315 behavior.

*This comment was generated by an AI coding assistant.*

File: e2e/cli/test_minimum_release_age_installed_resolution [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+# Make a remote lookup observable and fatal. Ordinary resolution should use
+# the latest installed match without calling this script, even with an
+# explicit release-age policy.
+cat >"$MISE_DATA_DIR/plugins/spy/bin/list-all" <<EOF
+#!/usr/bin/env bash
+touch "$marker"
+exit 1
+EOF
+rm -rf "$MISE_CACHE_DIR/spy"
```
[the remark this answers is not in the document]
Comment (jdx) [reply e2e/cli/test_minimum_release_age_installed_resolution:25]: Fixed in ff44ad595. The regression test now replaces both `list-all` and `latest-stable` with marker-touching failures, so either remote lookup path is detected. The focused E2E test and `mise run lint-fix` pass.

*This comment was generated by an AI coding assistant.*

File: docs/security.md [hunk 3]
```diff
@@ -44,6 +44,8 @@ minimum_release_age = "7d"  # only install versions released more than 7 days ag
 Supports relative durations (`7d`, `6mo`, `1y`) and absolute dates (`2024-06-01`). For most
 backends, this only affects fuzzy version resolution, such as `node@20` or `latest`.
 Explicitly pinned versions like `node@22.5.0` bypass the filter.
+Versions that are already installed also remain eligible: `minimum_release_age` limits remote
+version selection and does not make an installed version inactive.
```
[the remark this answers is not in the document]
Comment (jdx) [reply docs/security.md]: Fixed in ff44ad595. The documentation now limits installed-version eligibility to ordinary toolset resolution and explicitly notes that lockfile generation may re-check fuzzy installed matches against release metadata.

*This comment was generated by an AI coding assistant.*

File: e2e/cli/test_minimum_release_age_installed_resolution [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+cat >mise.toml <<EOF
+[settings]
+minimum_release_age = "2d"
+
+[tools]
+spy = "latest"
+EOF
+
+assert "mise bin-paths" "$MISE_DATA_DIR/installs/spy/latest/bin"
```
[the remark this answers is not in the document]
Comment (jdx) [reply e2e/cli/test_minimum_release_age_installed_resolution:38]: Clarified in 02b677e0a. The resolved version is concrete (`mise current spy` now explicitly asserts `2.0.0`), while `mise bin-paths` intentionally returns the fuzzy request's runtime symlink path (`spy/latest/bin`). Both assertions pass, and the remote-fetch marker remains untouched.

*This comment was generated by an AI coding assistant.*

File: src/cli/lock.rs [hunk 5]
```diff
@@ -107,12 +108,10 @@ impl Lock {
...
             ts_owned = monorepo_ts;
             &ts_owned
-        } else if before_date.is_some() {
+        } else {
             let builder = ToolsetBuilder::new().with_resolve_options(lock_resolve_options.clone());
             ts_owned = builder.build(&config).await?;
             &ts_owned
-        } else {
-            config.get_toolset().await?
         };
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/cli/lock.rs]: Fixed in 02b677e0a. A normal `mise lock` with no CLI, global, or per-tool release-age policy now reuses `config.get_toolset()` again. The lock-specific rebuild is limited to explicitly configured cutoffs, preserving the cached fast path while retaining release-age filtering for global and per-tool policies.

*This comment was generated by an AI coding assistant.*
